File: ide/jetbrains/src/main/kotlin/com/bc/plugin/services/BcService.kt
package com.bc.plugin.services

import com.intellij.execution.configurations.GeneralCommandLine
import com.intellij.execution.process.OSProcessHandler
import com.intellij.execution.process.ProcessOutput
import com.intellij.execution.util.ExecUtil
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import java.io.File

/**
 * Service for interacting with the bc CLI.
 */
@Service(Service.Level.PROJECT)
class BcService(private val project: Project) {
    private val logger = Logger.getInstance(BcService::class.java)

    private var bcPath: String = "bc"
    private var isWorkspace: Boolean = false

    init {
        detectWorkspace()
    }

    /**
     * Check if the project is a bc workspace (has .bc directory)
     */
    fun detectWorkspace(): Boolean {
        val basePath = project.basePath ?: return false
        val bcDir = File(basePath, ".bc")
        isWorkspace = bcDir.exists() && bcDir.isDirectory
        return isWorkspace
    }

    fun isWorkspace(): Boolean = isWorkspace

    /**
     * Execute a bc command and return the output
     */
    fun execute(vararg args: String): ProcessOutput? {
        if (!isWorkspace) return null

        return try {
            val commandLine = GeneralCommandLine(bcPath)
                .withParameters(*args)
                .withWorkDirectory(project.basePath)
                .withEnvironment("NO_COLOR", "1")

            ExecUtil.execAndGetOutput(commandLine, 30000)
        } catch (e: Exception) {
            logger.warn("Failed to execute bc command: ${args.joinToString(" ")}", e)
            null
        }
    }

    /**
     * Execute bc command and return stdout
     */
    fun executeAndGetOutput(vararg args: String): String? {
        val output = execute(*args) ?: return null
        return if (output.exitCode == 0) output.stdout else null
    }

    /**
     * Get workspace status
     */
    fun getStatus(): BcStatus? {
        val output = executeAndGetOutput("status", "--json") ?: return null
        return parseStatus(output)
    }

    /**
     * List all agents
     */
    fun listAgents(): List<Agent> {
        val output = executeAndGetOutput("agent", "list", "--json") ?: return emptyList()
        return parseAgents(output)
    }

    /**
     * List all channels
     */
    fun listChannels(): List<String> {
        val output = executeAndGetOutput("channel", "list") ?: return emptyList()
        return output.lines()
            .filter { it.isNotBlank() }
            .map { it.trim() }
    }

    /**
     * Get channel history
     */
    fun getChannelHistory(channel: String, limit: Int = 20): List<ChannelMessage> {
        val output = executeAndGetOutput("channel", "history", channel, "--limit", limit.toString())
            ?: return emptyList()
        return parseChannelHistory(output)
    }

    /**
     * Send message to channel
     */
    fun sendToChannel(channel: String, message: String): Boolean {
        val output = execute("channel", "send", channel, message)
        return output?.exitCode == 0
    }

    /**
     * Get recent logs
     */
    fun getLogs(limit: Int = 50): String {
        return executeAndGetOutput("logs", "--tail", limit.toString()) ?: ""
    }

    fun setBcPath(path: String) {
        bcPath = path
    }

    fun getBcPath(): String = bcPath

    // Data classes
    data class BcStatus(
        val workspace: String,
        val agentCount: Int,
        val activeCount: Int,
        val workingCount: Int
    )

    data class Agent(
        val name: String,
        val role: String,
        val state: String,
        val uptime: String,
        val task: String
    )

    data class ChannelMessage(
        val timestamp: String,
        val sender: String,
        val message: String
    )

    // Parsers
    private fun parseStatus(json: String): BcStatus? {
        // Simple parsing - in production would use kotlinx.serialization
        return try {
            val workspace = Regex(""""workspace":\s*"([^"]+)"""").find(json)?.groupValues?.get(1) ?: "unknown"
            val agentCount = Regex(""""agent_count":\s*(\d+)""").find(json)?.groupValues?.get(1)?.toInt() ?: 0
            val activeCount = Regex(""""active_count":\s*(\d+)""").find(json)?.groupValues?.get(1)?.toInt() ?: 0
            val workingCount = Regex(""""working_count":\s*(\d+)""").find(json)?.groupValues?.get(1)?.toInt() ?: 0
            BcStatus(workspace, agentCount, activeCount, workingCount)
        } catch (e: Exception) {
            null
        }
    }

    private fun parseAgents(json: String): List<Agent> {
        // Simple line-based parsing for table output
        return json.lines()
            .filter { it.contains("engineer") || it.contains("manager") || it.contains("root") }
            .mapNotNull { line ->
                val parts = line.split(Regex("\\s{2,}")).map { it.trim() }
                if (parts.size >= 4) {
                    Agent(
                        name = parts[0],
                        role = parts[1],
                        state = parts[2],
                        uptime = parts.getOrElse(3) { "-" },
                        task = parts.getOrElse(4) { "" }
                    )
                } else null
            }
    }

    private fun parseChannelHistory(output: String): List<ChannelMessage> {
        return output.lines()
            .filter { it.contains("[") && it.contains("]") }
            .mapNotNull { line ->
                val match = Regex("""\[(\d+)\]\s*\[([^\]]+)\]\s*([^:]+):\s*(.*)""").find(line)
                if (match != null) {
                    ChannelMessage(
                        timestamp = match.groupValues[2],
                        sender = match.groupValues[3].trim(),
                        message = match.groupValues[4].trim()
                    )
                } else null
            }
    }

    companion object {
        fun getInstance(project: Project): BcService {
            return project.getService(BcService::class.java)
        }
    }
}

File: ide/vscode/src/services/bcService.ts
import { execSync, exec } from 'child_process';
import * as path from 'path';
import * as fs from 'fs';
import * as vscode from 'vscode';

export interface BcStatus {
    workspace: string;
    agentCount: number;
    activeCount: number;
    workingCount: number;
}

export interface Agent {
    name: string;
    role: string;
    state: string;
    uptime: string;
    task: string;
}

export interface ChannelMessage {
    timestamp: string;
    sender: string;
    message: string;
}

export class BcService {
    private workspacePath: string;
    private bcPath: string;
    private _isWorkspace: boolean;

    constructor(workspacePath: string) {
        this.workspacePath = workspacePath;
        this.bcPath = vscode.workspace.getConfiguration('bc').get('binaryPath', 'bc');
        this._isWorkspace = this.detectWorkspace();
    }

    private detectWorkspace(): boolean {
        const bcDir = path.join(this.workspacePath, '.bc');
        return fs.existsSync(bcDir) && fs.statSync(bcDir).isDirectory();
    }

    isWorkspace(): boolean {
        return this._isWorkspace;
    }

    async execute(...args: string[]): Promise<string | null> {
        if (!this._isWorkspace) {
            return null;
        }

        return new Promise((resolve) => {
            const cmd = `${this.bcPath} ${args.join(' ')}`;
            exec(cmd, {
                cwd: this.workspacePath,
                env: { ...process.env, NO_COLOR: '1' },
                timeout: 30000
            }, (error, stdout) => {
                if (error) {
                    console.warn(`bc command failed: ${args.join(' ')}`, error);
                    resolve(null);
                } else {
                    resolve(stdout);
                }
            });
        });
    }

    async getStatus(): Promise<BcStatus | null> {
        const output = await this.execute('status', '--json');
        if (!output) {
            return null;
        }

        try {
            const data = JSON.parse(output);
            return {
                workspace: data.workspace || 'unknown',
                agentCount: data.agent_count || 0,
                activeCount: data.active_count || 0,
                workingCount: data.working_count || 0
            };
        } catch {
            return null;
        }
    }

    async listAgents(): Promise<Agent[]> {
        const output = await this.execute('agent', 'list', '--json');
        if (!output) {
            return [];
        }

        try {
            const data = JSON.parse(output);
            if (Array.isArray(data)) {
                return data.map((a: Record<string, unknown>) => ({
                    name: String(a.name || ''),
                    role: String(a.role || ''),
                    state: String(a.state || ''),
                    uptime: String(a.uptime || '-'),
                    task: String(a.task || '')
                }));
            }
        } catch {
            // Parse table output as fallback
            return this.parseAgentTable(output);
        }
        return [];
    }

    private parseAgentTable(output: string): Agent[] {
        return output.split('\n')
            .filter(line => line.includes('engineer') || line.includes('manager') || line.includes('root'))
            .map(line => {
                const parts = line.split(/\s{2,}/).map(p => p.trim());
                if (parts.length >= 4) {
                    return {
                        name: parts[0],
                        role: parts[1],
                        state: parts[2],
                        uptime: parts[3] || '-',
                        task: parts[4] || ''
                    };
                }
                return null;
            })
            .filter((a): a is Agent => a !== null);
    }

    async listChannels(): Promise<string[]> {
        const output = await this.execute('channel', 'list');
        if (!output) {
            return [];
        }

        return output.split('\n')
            .map(line => line.trim())
            .filter(line => line.length > 0 && !line.startsWith('CHANNEL'));
    }

    async getChannelHistory(channel: string, limit: number = 20): Promise<ChannelMessage[]> {
        const output = await this.execute('channel', 'history', channel, '--limit', String(limit));
        if (!output) {
            return [];
        }

        const messages: ChannelMessage[] = [];
        const regex = /\[(\d+)\]\s*\[([^\]]+)\]\s*([^:]+):\s*(.*)/;

        for (const line of output.split('\n')) {
            const match = regex.exec(line);
            if (match) {
                messages.push({
                    timestamp: match[2],
                    sender: match[3].trim(),
                    message: match[4].trim()
                });
            }
        }

        return messages;
    }

    async sendToChannel(channel: string, message: string): Promise<boolean> {
        const output = await this.execute('channel', 'send', channel, message);
        return output !== null;
    }

    async getLogs(limit: number = 50): Promise<string> {
        const output = await this.execute('logs', '--tail', String(limit));
        return output || '';
    }
}
